unixfs/hamt/hamt.go

```diff
@@ -399,30 +492,33 @@ func (ds *Shard) EnumLinks(ctx context.Context) ([]*ipld.Link, error) {
 }
 
 // ForEachLink walks the Shard and calls the given function.
-func (ds *Shard) ForEachLink(ctx context.Context, f func(*ipld.Link) error) error {
-	return ds.walkTrie(ctx, func(sv *shardValue) error {
+func (ds *Shard) ForEachLink(ctx0 context.Context, f func(*ipld.Link) error) error {
+	ctx, cancel := context.WithCancel(ctx0)
+	fetcher := startFetcher(ctx, ds.dserv)
+	err := ds.walkTrie(ctx, func(sv *shardValue) error {
 		lnk := sv.val
 		lnk.Name = sv.key
 
 		return f(lnk)
-	})
+	}, fetcher)
+	cancel()
```

Comment:
why not just defer cancel?

Reply:
Fixed.  (I didn't initially because I didn't see the need, but I guess there is no harm in doing it)